primitives/native-token-management/src/lib.rs

```diff
@@ -0,0 +1,127 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(feature = "std")]
+pub use inherent_provider::*;
+
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use sidechain_domain::*;
+use sp_inherents::*;
+use sp_runtime::{
+	scale_info::TypeInfo,
+	traits::{Block as BlockT, Header, Zero},
+};
+
+#[cfg(test)]
+mod tests;
+
+pub const INHERENT_IDENTIFIER: InherentIdentifier = *b"ntreserv";
+
+#[derive(Default, Debug, Clone, PartialEq, Eq, TypeInfo, Encode, Decode, MaxEncodedLen)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct MainChainScripts {
+	pub native_token_policy: PolicyId,
+	pub native_token_asset_name: AssetName,
+	pub illiquid_supply_address: MainchainAddress,
+}
+
+sp_api::decl_runtime_apis! {
+	pub trait NativeTokenManagementApi {
+		fn get_main_chain_scripts() -> MainChainScripts;
+	}
+}
+
+#[derive(Decode, Encode)]
+pub struct TokenTransferData {
+	pub token_amount: NativeTokenAmount,
+}
+
+#[cfg(feature = "std")]
+mod inherent_provider {
+	use super::*;
+	use main_chain_follower_api::{DataSourceError, NativeTokenManagementDataSource};
+	use sidechain_mc_hash::McHashInherentDigest;
+	use sp_api::{ApiError, ProvideRuntimeApi};
+	use sp_blockchain::HeaderBackend;
+	use std::sync::Arc;
+
+	pub struct NativeTokenManagementInherentDataProvider {
+		pub token_amount: NativeTokenAmount,
+	}
+
+	#[derive(thiserror::Error, sp_runtime::RuntimeDebug)]
+	pub enum IDPCreationError {
+		#[error("Failed to read native token data from data source: {0:?}")]
+		DataSourceError(DataSourceError),
```

Comment:
I think that you can use a @from annotation here instead of manually adding From implementation below.